### PR TITLE
Improve coverage

### PR DESCRIPTION
**Pull request by krlmlr**
(> 98% now) and fix a few minor problems.

- `vec_assert()` supports arbitrary expressions
- `vec_count()` now works in corner cases and for data frames with `sort = "key"`
- Casting `list_of` to `list` is never lossy
- Shortcut `"prtl"` for `vctrs_partial_frame` class
- Fix casting lists to records
- `vec_slice()` now appropriately recycles indexes of type `logical`
- `vec_names<-()` never sets row names on data frames
- Removed specializations of `vec_type2.unspecified`

### CODE REVIEW

File: R/assert.R [hunk 1]
```diff
@@ -6,7 +6,7 @@
 #' @return Either an error, or `x`, invisibly.
 #' @export
 vec_assert <- function(x, ptype = NULL, size = NULL) {
-  x_name <- as.character(ensym(x))
+  x_name <- as_label(x)
```
Comment (hadley) [inline R/assert.R]: Does this need a bump to required rlang version? (IIRC it's fairly new)

File: R/conditions.R [hunk 2]
```diff
@@ -13,14 +13,7 @@
...
 
-stop_incompatible <- function(x, y, details = NULL, ..., message = NULL, .subclass = NULL) {
-  if (is.null(message)) {
-    message <- glue::glue("Incompatible inputs: <{vec_ptype_full(x)}> and <{vec_ptype_full(y)}>")
-    if (!is.null(details)) {
-      message <- paste0(message, "\n", details)
-    }
-  }
-
+stop_incompatible <- function(x, y, message, details = NULL, ..., .subclass = NULL) {
```
Comment (hadley) [inline R/conditions.R]: I think it's bad for the parent "constructor" to have a different signature to the children.

File: R/dictionary.R [hunk 3]
```diff
@@ -44,15 +44,16 @@ vec_count <- function(x, sort = c("count", "key", "location", "none")) {
...
+  df <- data.frame(key = rep_along(kv$val, NA), count = kv$val)
   df$key <- vec_slice(x, kv$key) # might be a dataframe
 
   if (sort == "none")
     return(df)
 
   idx <- switch(sort,
     location = order(kv$key),
-    key = order(df$key),
+    key = order_proxy(vec_proxy_compare(df$key)),
```
Comment (hadley) [inline R/dictionary.R]: Why did this change?

File: R/list_of.R [hunk 4]
```diff
@@ -159,13 +159,16 @@ vec_cast.vctrs_list_of <- function(x, to) {
...
+  # but the locations refer to the inner vectors,
+  # and the cast fails if all (vector) elements in a single (list) element
   as_list_of(x, .ptype = attr(to, "ptype"))
 }
 
 #' @export
 #' @method vec_cast.list vctrs_list_of
 vec_cast.list.vctrs_list_of <- function(x, to) {
-  warn_lossy_cast(x, to)
+  # Casting list_of to list is never lossy
```
Comment (hadley) [inline R/list_of.R]: Isn't it? We're losing the type information about the contents, which seems lossy to me.

File: R/rcrd.R [hunk 5]
```diff
@@ -74,16 +76,33 @@ obj_str_data.vctrs_rcrd <- function(x, ...) {
...
-#' @method vec_cast.vctrs_rcrd list
+#' @method vec_cast.vctrs_rcrd vctrs_rcrd
 #' @export
-vec_cast.vctrs_rcrd.list <- function(x, to) {
-  vec_list_cast(x, to)
+vec_cast.vctrs_rcrd.vctrs_rcrd <- function(x, to) {
+  if (n_fields(x) != n_fields(to)) {
+    stop_incompatible_cast(x, to)
+  }
+  if (any(sort(fields(x)) != sort(fields(to)))) {
```
Comment (hadley) [inline R/rcrd.R]: Could we use `setequal()` here?

File: R/size.R [hunk 6]
```diff
@@ -99,9 +99,7 @@ vec_size2 <- function(nx, ny) {
 #'
 #' vec_slice(mtcars, 1:3)
 vec_slice <- function(x, i) {
-  if (is.logical(i)) {
-    i <- which(i)
-  }
+  i <- recycle_logical_index(i, x)
   stopifnot(is.integer(i) || is.character(i))
```
Comment (hadley) [inline R/size.R]: Shouldn't the `stopifnot()` move inside of `recycle_logical_index()`?

File: R/size.R [hunk 7]
```diff
@@ -185,7 +189,7 @@ vec_names <- function(x) {
 `vec_names<-` <- function(x, value) {
   if (vec_dims(x) == 1) {
     names(x) <- value
-  } else {
+  } else if (!is.data.frame(x)) {
```
Comment (hadley) [inline R/size.R]: It might be clearer to parallel exactly the same `if` blocks as above.

File: R/unspecified.R [hunk 8]
```diff
@@ -55,76 +55,39 @@ vec_type_finalise.vctrs_unspecified <- function(x) {
...
-#' @method vec_type2 vctrs_unspecified
 #' @export
-vec_type2.vctrs_unspecified <- function(x, y) UseMethod("vec_type2.vctrs_unspecified", y)
-#' @method vec_type2.vctrs_unspecified default
-#' @export
-vec_type2.vctrs_unspecified.default <- function(x, y) y
-#' @method vec_type2.vctrs_unspecified vctrs_unspecified
-#' @export
-vec_type2.vctrs_unspecified.vctrs_unspecified <- function(x, y) x
+vec_type2.vctrs_unspecified <- function(x, y) vec_slice(y, 0L)
```
Comment (hadley) [inline R/unspecified.R]: I think these should all call `vec_type()` instead of `vec_slice(y, 0L)`

File: R/vctr.R [hunk 9]
```diff
@@ -387,12 +387,18 @@ xtfrm.vctrs_vctr <- function(x) {
...
+  if (is_integer(proxy) || is_double(proxy)) {
     proxy
-  } else if (is_character(proxy) || is_logical(proxy)) {
-    order(order(proxy))
+  } else {
+    order(order_proxy(proxy))
+  }
+}
+
+order_proxy <- function(proxy) {
```
Comment (hadley) [inline R/vctr.R]: I don't understand why you extracted this out.

File: tests/testthat/test-assert.R [hunk 10]
```diff
@@ -0,0 +1,32 @@
+context("test-assert")
+
+test_that("basic assert is a no-op", {
```
Comment (hadley) [inline tests/testthat/test-assert.R]: is idempotent?

File: tests/testthat/test-assert.R [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+  expect_identical(vec_assert(1:4), 1:4)
+})
+
+test_that("asserting ptype", {
+  expect_error(vec_assert(new_vctr(1:4), new_vctr(integer())), NA)
+
+  # Is this the correct error message?
+  expect_error(
+    vec_assert(new_vctr(1:4), new_vctr(double())),
+    "must be <vctrs_vc>, not <vctrs_vc>.",
```
Comment (hadley) [inline tests/testthat/test-assert.R]: Should we be checking for a class here?

File: tests/testthat/test-cast-list.R [hunk 12]
```diff
@@ -15,3 +15,10 @@ test_that("elements of length >1 are truncated with warning", {
   out <- expect_warning(vec_list_cast(x, double()), class = "warn_lossy_cast")
   expect_equal(out, c(1, 2, 3))
 })
+
+test_that("error if all casts are lossy", {
+  # Do we really want to err out here? Perhaps give a different warning?
```
Comment (hadley) [inline tests/testthat/test-cast-list.R]: Can you file an issue? It's possible we need a useful friendly form and a stricter programming form.

File: tests/testthat/test-dictionary.R [hunk 13]
```diff
@@ -8,6 +8,11 @@ test_that("vec_count counts number observations", {
   expect_equal(x, data.frame(key = 1:3, count = 1:3))
 })
 
+test_that("vec_count works in corner cases", {
```
Comment (hadley) [inline tests/testthat/test-dictionary.R]:
```suggestion
test_that("vec_count works for zero-length input", {
```

File: tests/testthat/test-rcrd.R [hunk 14]
```diff
@@ -59,6 +112,16 @@ test_that("no attributes", {
   expect_error(new_rcrd(x), "no attributes")
 })
 
+test_that("subset assignment", {
```
Comment (hadley) [inline tests/testthat/test-rcrd.R]:
```suggestion
test_that("subset assignment throws error", {
```

File: tests/testthat/test-rcrd.R [hunk 15]
```diff
@@ -59,6 +112,16 @@ test_that("no attributes", {
...
+  x <- new_rcrd(list(x = 1))
+  expect_error(
+    x$y <- 2,
+    "subset assignment with $",
```
Comment (hadley) [inline tests/testthat/test-rcrd.R]: I don't think you should test the text here

File: tests/testthat/test-sclr.R [hunk 16]
```diff
@@ -31,6 +31,14 @@ test_that("as.list strips attributes apart from names", {
   expect_equal(attributes(y), list(names = names(x)))
 })
 
+test_that("as.data.frame works", {
```
Comment (hadley) [inline tests/testthat/test-sclr.R]: I'm not 100% certain this is correct behaviour. Can you please file an issue instead of making a test?

File: R/dictionary.R (hunk 3 above)
Comment (krlmlr) [reply R/dictionary.R]: > vec_count() now works in corner cases and for data frames with sort = "key"

``` r
withr::with_options(list(max.print = 100), print(vctrs::vec_count(
  iris,
  "key"
)))
#>    key.Sepal.Length key.Sepal.Width key.Petal.Length key.Petal.Width
#> 1                NA              NA               NA              NA
#> 2                NA              NA               NA              NA
#> 3                NA              NA               NA              NA
#> 4                NA              NA               NA              NA
#> 5                NA              NA               NA              NA
#> 6                NA              NA               NA              NA
#> 7                NA              NA               NA              NA
#> 8                NA              NA               NA              NA
#> 9                NA              NA               NA              NA
#> 10               NA              NA               NA              NA
#> 11               NA              NA               NA              NA
#> 12               NA              NA               NA              NA
#> 13               NA              NA               NA              NA
#> 14               NA              NA               NA              NA
#> 15               NA              NA               NA              NA
#> 16               NA              NA               NA              NA
#>    key.Species count
#> 1         <NA>    NA
#> 2         <NA>    NA
#> 3         <NA>    NA
#> 4         <NA>    NA
#> 5         <NA>    NA
#> 6         <NA>    NA
#> 7         <NA>    NA
#> 8         <NA>    NA
#> 9         <NA>    NA
#> 10        <NA>    NA
#> 11        <NA>    NA
#> 12        <NA>    NA
#> 13        <NA>    NA
#> 14        <NA>    NA
#> 15        <NA>    NA
#> 16        <NA>    NA
#>  [ reached getOption("max.print") -- omitted 34 rows ]
#>  [ reached 'max' / getOption("max.print") -- omitted 695 rows ]
```

<sup>Created on 2019-01-25 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

File: R/list_of.R (hunk 4 above)
Comment (krlmlr) [reply R/list_of.R]: Is this a lossy cast?

``` r
vctrs::vec_cast(1:3, list())
#> [[1]]
#> [1] 1
#> 
#> [[2]]
#> [1] 2
#> 
#> [[3]]
#> [1] 3
```

<sup>Created on 2019-01-25 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

I'd argue that a cast is lossy only if the data is changed, and an upcast (to a more general type) is never lossy.

File: tests/testthat/test-sclr.R (hunk 16 above)
Comment (krlmlr) [reply tests/testthat/test-sclr.R]: Done.

File: R/vctr.R (hunk 9 above)
Comment (krlmlr) [reply R/vctr.R]: So that I can do `order_proxy(vec_proxy_compare(df$key))` . Do you see a better way?

File: R/unspecified.R (hunk 8 above)
Comment (krlmlr) [reply R/unspecified.R]: Changed two more instances of the same pattern.

File: tests/testthat/test-assert.R (hunk 11 above)
Comment (krlmlr) [reply tests/testthat/test-assert.R]: Filed a new issue.

**Comment by krlmlr** [comment]
Addressed your comments, 3 1/2 questions remain.

File: R/list_of.R (hunk 4 above)
Comment (hadley) [reply R/list_of.R]: Oh got it.

File: DESCRIPTION [hunk 17]
```diff
@@ -1,35 +1,38 @@
...
 BugReports: https://github.com/r-lib/vctrs/issues
 Depends: 
     R (>= 3.1)
 Imports: 
-    backports,
-    digest,
-    glue,
-    rlang,
-    zeallot
+    backports (>= 1.1.3),
```
Comment (hadley) [inline DESCRIPTION]: Can you please just bump the dependency for rlang, not everything?

File: R/conditions.R [hunk 18]
```diff
@@ -13,14 +13,7 @@
...
 
-stop_incompatible <- function(x, y, details = NULL, ..., message = NULL, .subclass = NULL) {
-  if (is.null(message)) {
-    message <- glue::glue("Incompatible inputs: <{vec_ptype_full(x)}> and <{vec_ptype_full(y)}>")
-    if (!is.null(details)) {
-      message <- paste0(message, "\n", details)
-    }
-  }
-
+stop_incompatible <- function(x, y, details = NULL, ..., message, .subclass = NULL) {
```
Comment (hadley) [inline R/conditions.R]: Now it's weird not to have an default for `message`, so I think you should leave the previous code.

**Comment by krlmlr** [comment]
Done, thanks for the quick review.
